pkg/kubelet/preemption/preemption.go

```diff
@@ -0,0 +1,287 @@
+type CriticalPodAdmissionHandler struct {
+	getPodsFunc eviction.ActivePodsFunc
+	killPodFunc eviction.KillPodFunc
+	recorder    record.EventRecorder
+}
+
+var _ lifecycle.AdmissionFailureHandler = &CriticalPodAdmissionHandler{}
+
+func NewCriticalPodAdmissionHandler(getPodsFunc eviction.ActivePodsFunc, killPodFunc eviction.KillPodFunc, recorder record.EventRecorder) *CriticalPodAdmissionHandler {
+	return &CriticalPodAdmissionHandler{
+		getPodsFunc: getPodsFunc,
+		killPodFunc: killPodFunc,
+		recorder:    recorder,
+	}
+}
+
+// HandleAdmissionFailure gracefully handles admission rejection, and, in some cases,
+// to allow admission of the pod despite its previous failure.
+func (c *CriticalPodAdmissionHandler) HandleAdmissionFailure(pod *v1.Pod, failureReasons []algorithm.PredicateFailureReason) (bool, []algorithm.PredicateFailureReason, error) {
+	if !kubetypes.IsCriticalPod(pod) || !utilfeature.DefaultFeatureGate.Enabled(features.ExperimentalCriticalPodAnnotation) {
+		return false, failureReasons, nil
+	}
+	// InsufficientResourceError is not a reason to reject a critical pod.
+	// Instead of rejecting, we free up resources to admit it, if no other reasons for rejection exist.
+	nonResourceReasons := []algorithm.PredicateFailureReason{}
+	resourceReasons := []*admissionRequirement{}
+	for _, reason := range failureReasons {
+		if r, ok := reason.(*predicates.InsufficientResourceError); ok {
+			resourceReasons = append(resourceReasons, &admissionRequirement{
+				resourceName: r.ResourceName,
+				quantity:     r.GetInsufficientAmount(),
+			})
+		} else {
+			nonResourceReasons = append(nonResourceReasons, reason)
+		}
+	}
+	if len(nonResourceReasons) > 0 {
+		// Return only reasons that are not resource related, since critical pods cannot fail admission for resource reasons.
+		return false, nonResourceReasons, nil
+	}
+	err := c.evictPodsToFreeRequests(admissionRequirementList(resourceReasons))
+	// if no error is returned, preemption succeeded and the pod is safe to admit.
+	return err == nil, nil, err
+}
+
+// freeRequests takes a list of insufficient resources, and attempts to free them by evicting pods
+// based on requests.  For example, if the only insufficient resource is 200Mb of memory, this function could
+// evict a pod with request=250Mb.
+func (c *CriticalPodAdmissionHandler) evictPodsToFreeRequests(insufficientResources admissionRequirementList) error {
+	podsToPreempt, err := getPodsToPreempt(c.getPodsFunc(), insufficientResources)
+	if err != nil {
+		return fmt.Errorf("preemption: error finding a set of pods to preempt: %v", err)
+	}
+	glog.Infof("preemption: attempting to evict pods %v, in order to free up resources: %s", podsToPreempt, insufficientResources.toString())
+	for _, pod := range podsToPreempt {
+		status := v1.PodStatus{
+			Phase:   v1.PodFailed,
+			Message: message,
+			Reason:  reason,
+		}
+		// record that we are evicting the pod
+		c.recorder.Eventf(pod, v1.EventTypeWarning, reason, message)
```

Comment:
I was basing this off of the eviction manager code, which doesnt address this question in the original PR when it was added #21274.  I suspect that we want to avoid the scenario where we kill a pod, and the kubelet restarts before we record the event.  This would mysteriously cause the pod to disappear from the node without an event.